### ISSUE TITLE
Feature Request - Save as markdown file to local filesystem

### ISSUE THREAD

**Issue by EphemeralDev** [quillpad/quillpad#26]
Similar feature requested in  [quillnote](https://github.com/msoultanidis/quillnote/issues/9). Would be nice to be able to sync plaintext files using syncthing or similar software.

**Comment by jsixface** [comment]
This is one of the features I wanted to have in version 1.5. The way I see it, we'll use Android's Storage access framework to implement the functionality to select the directory under which we want to sync the markdown files to. That directory can be from a cloud provider like Google Drive, Nextcloud or even Dropbox and we'll have the ability to sync to any cloud provider. 

I'll later break this down to smaller issue tickets and tag is to the milestone. 

In a later version, we could add option to encrypt the stored files so that they stay private.

**Comment by jsixface** [comment]
Updated the README with milestones to reflect this

**Comment by trymeouteh** [comment]
Would be great to use this app with Syncthing, and to be able to have notes encrypted just like Joplin.

**Comment by Terkyz** [comment]
> Would be great to use this app with Syncthing, and to be able to have notes encrypted just like Joplin.

Wouldn't it be better if the app had an option to automatically sync with the desktop version if they are in the same network and with some verification method, like an QR code? I have used Joplin for a long time and getting the desktop and phone version to sync with Syncthing was a headache, if something like automatically local sync was integrated into Quillpad it would be better in my opinion.

**Comment by jsixface** [comment]
**Update:** I just found out that Google drive and Microsoft OneDrive apps don't support sharing a folder to other apps using the Storage API. Nextcloud and pCloud do, so that's better

**Comment by NVolcz** [comment]
I am also interested in having the ability to save notes as markdown files in a specific directory. Have you hade time to create separate issues for the tasks? I am interested in helping out.

**Comment by nWestie** [comment]
same here, would like this feature to exist and am interested in working on it. From what I understand, what it should do, if enabled by the user, is:
- save all notes to the user specified folder.
  - does this include lists? Can they be fully represented as Markdown?
- If the contents of a note are updated, Quillpad will reflect these changes

Thoughts/Questions:
- Can Notes/Lists be fully represented in markdown without losing any content/properties? I think this would be required if we want Quillpad to be able to follow changes made in the filesystem.
- How are notebooks, trash, archive represented? Perhaps subfolders?
  - Another option would be to have a table in the markdown with notebook/color/trash, I'm imagining something like [Jekyll](https://jekyllrb.com/docs/front-matter/) does but it doesn't have to be that.
- if the goal is interoperability with other note software, is there anything that is required to make this work well? 
- I'm still getting up to speed with android, but I'd imagine there's a way to get a lock on a file? That way we shouldn't have to worry about Quillpad being out of sync with the file system, and any conflicts would be handled by the user or syncing software.

As a first step, it would be nice to have backups happen regularly, instead of needing to trigger them manually.

**Comment by NVolcz** [comment]
I understand that the original intention was to use the Android Storage API but it sounds that it is severely limited due to how other apps use it. It is likely that another library or some custom code is needed.

Regarding storing metadata about the notes, I see three places where we can store metadata:
1. The directory structure - nice for some use cases but not likely to be compatible with other softwares and the user can also have opinions how to structure the storage.
2. Front matter - makes the metadata embedded into the notes which is nice when you want to be able to move it with the notes but looks terrible if not supported by the editor.
3. Separate storage - looks nice, flexible but does not move together with the data.

**Comment by jsixface** [comment]
@nWestie Thank you for your interest in working on it. 

I have made significant progress on this and the code is in `local-storage` branch of this repo. https://github.com/quillpad/quillpad/tree/local-storage
It is still a development branch and I have gotten it up to a point that is able to store the notes in `.md` files and use Android storage API. The hiccup here is the storage API performance when that storage is backed by a cloud directory provider like Nextcloud or pCloud. 

Database: Quillpad stores all the notes, tasks, and sync information in a Room database. The notes in the db is synced with Nextcloud and now the content is also synced with file system. I made some database migration changes but that has some conflicts with recent Pull request for storing the collapsed version of notes in the UI. 


Regarding Android storage API - I would like to keep the storage API as the file system access mechanism since that's the recommended file system access by Android for use cases like this. 

@NVolcz & @nWestie Take a look at the changes on that branch and you'll have a better understanding of the changes needed. I am planning to fix the performance and create a PR for this. If you like to work on this branch, let me know in the Matrix chat room, and I can open it up to you.

**Comment by the-all** [comment]
I would also love this feature just as an option for note export. Sometimes, I want to send a note to people over a messenger app, but as .txt or .md attachment instead of inline text. As it stands, I can either use the share menu to send out just the text content, or the export menu which gives a .zip file. The former might clog up the flow of a conversation, the latter opens up an additional step in opening.

**Comment by RustoMCSpit** [comment]
any update?

**Comment by RustoMCSpit** [comment]
any update??

**Comment by RayBB** [comment]
Hi @jsixface and friends,
I wanted to say thanks so much for developing this app and sharing it with the world.

I love the UX of Google Keep but want to keep my notes as markdown and use them in different places. Your app is the perfect UI already! Currently I use Obsidian for notes on my laptop and syncthing to send it to my phone. The only problem is that it is a bit slow and clunky on my phone and I want a Google Keep style interface to see recent notes. I've been thinking about cooking up (with chatGPT and Flutter) something simple like that when I found your app. It pretty much has the perfect UX but just lacks the ability to point it to the folder where my obsidian notes are sync'd.

Unfortunately, I have basically no android experience so jumping into this code base isn't going to happen anytime soon. 

However, I want to offer a small $50 donation (to whoever merges) as a token of my appreciation if this feature is released :)
I know it isn't much but I'm just a student now. I think it will help many people so I want to encourage it.

My conditions are:
1. I can select a folder where my notes are stored
2. It will show notes from that folder and all subfolders
3. It is merged before June 30, 2024 (because I can't put promises out forever!)

Anyway, I hope all is well and you're still enjoying working on this project :)

**Comment by RustoMCSpit** [comment]
is there an update on this?

**Comment by RustoMCSpit** [comment]
@jsixface any update?

**Comment by RustoMCSpit** [comment]
@EphemeralDev @jsixface can one of you change the title of this to Save as Markdown text to local file system? I assume there'd be an option?

**Comment by EphemeralDev** [comment]
> @EphemeralDev @jsixface can one of you change the title of this to Save as Markdown text to local file system? I assume there'd be an option?

I can but markdown is plaintext, i just went with the less specific language. I can alter it to specify markdown if you think it clarifies the feature request. Some desktop editors may not launch relevant LSPs for txt files (unless specified) so it makes sense to me.

**Comment by RustoMCSpit** [comment]
> > @EphemeralDev @jsixface can one of you change the title of this to Save as Markdown text to local file system? I assume there'd be an option?
> 
> I can but markdown is plaintext, i just went with the less specific language. I can alter it to specify markdown if you think it clarifies the feature request. Some desktop editors may not launch relevant LSPs for txt files (unless specified) so it makes sense to me.

yeah alter it please

**Comment by RustoMCSpit** [comment]
@jsixface why was this unpinned?

**Comment by RustoMCSpit** [comment]
> > I have made significant progress on this and the code is in `local-storage` branch of this repo. https://github.com/quillpad/quillpad/tree/local-storage It is still a development branch and I have gotten it up to a point that is able to store the notes in `.md` files and use Android storage API. The hiccup here is the storage API performance when that storage is backed by a cloud directory provider like Nextcloud or pCloud.
> 
> thats the use case for the vast majority of users, can this get merged and have a warning it doesnt yet work for Nextcloud or pCloud or as a setting???

@jsixface sorry for the spam but theres been radio silence on your end. why not merge this branch as it is?

**Comment by RustoMCSpit** [comment]
to my understanding, beyond the cloud syncing it is fully functional. that serves 90% of us. i mean, even having it as a setting with a warning it's experimental would suffice. i have 1000s of notes on this app and need to get them off asap. thank you for all that you do but if im not missing something can you please merge the branch

**Comment by jsixface** [comment]
The reason for the delay is because the sync mechanism is complicated and there's major change to the syncing. I couldn't work on it lately. I wouldn't risk releasing half baked feature that could potentially mess up users notes. It did mess up my test notes.

**Comment by RustoMCSpit** [comment]
> The reason for the delay is because the sync mechanism is complicated and there's major change to the syncing. I couldn't work on it lately. I wouldn't risk releasing half baked feature that could potentially mess up users notes. It did mess up my test notes.

does this ONLY apply to sync notes? you could make it so the feature will not turn on if a user is syncing notes

**Comment by ConstantMan** [comment]
Most of us i think , just want a way to export and manage all the notes in a readable format like .md or even .txt  .

**Comment by RustoMCSpit** [comment]
> Most of us i think , just want a way to export and manage all the notes in a readable format like .md or even .txt .

exactly, the extreme vast majority of users are being held back for a niche use case which could be resolved by simply having an auto turn off of nextcloud if activated (as well as auto-creating a backup before the feature is turned on)

**Comment by RustoMCSpit** [comment]
im not trying to be rude, but i really amnt following as the branch is right there to be merged

**Comment by ConstantMan** [comment]
> > Most of us i think , just want a way to export and manage all the notes in a readable format like .md or even .txt .
> 
> exactly, the extreme vast majority of users are being held back for a niche use case which could be resolved by simply having an auto turn off of nextcloud if activated (as well as auto-creating a backup before the feature is turned on)

It's not a coincidence that this issue has the most comments

**Comment by kajdo** [comment]
my suggestion would be to have an option "within" the sync menu - so the user 
- *either* chooses nextcloud
- *or* chooses "save on local storage" (where he can chose the destination/source directory based on what SAF is offering)

this would provide users who want to setup an decoupled sync (rclone, webdav, sftp, git, ....) a nice way to do so and would not mess with the problems you face with the nextcloud/pcloud sync

just seperate this feature completely and make the "offline" functionality storing the .md files in a SAF provided directory

in my case I would use it together with my obsidian vault where I would put in a dedicated "QuickNotes" folder for Quillnotes. All the vaults data is in the "Downloads" directory which is exposed by SAF and I would take care about the sync however I sync my vault

the app looks beautiful and i think this would fix a bunch of issues mentioned in this thread

**Comment by regs01** [comment]
> * _either_ chooses nextcloud

Nextcloud is creating virtual drive in the system. So you can just choose it. So as many other cloud apps, and so as other many other aps like Termux.

**Comment by RustoMCSpit** [comment]
> Hi @jsixface and friends, I wanted to say thanks so much for developing this app and sharing it with the world.
> 
> I love the UX of Google Keep but want to keep my notes as markdown and use them in different places. Your app is the perfect UI already! Currently I use Obsidian for notes on my laptop and syncthing to send it to my phone. The only problem is that it is a bit slow and clunky on my phone and I want a Google Keep style interface to see recent notes. I've been thinking about cooking up (with chatGPT and Flutter) something simple like that when I found your app. It pretty much has the perfect UX but just lacks the ability to point it to the folder where my obsidian notes are sync'd.
> 
> Unfortunately, I have basically no android experience so jumping into this code base isn't going to happen anytime soon.
> 
> However, I want to offer a small $50 donation (to whoever merges) as a token of my appreciation if this feature is released :) I know it isn't much but I'm just a student now. I think it will help many people so I want to encourage it.
> 
> My conditions are:
> 
> 1. I can select a folder where my notes are stored
> 2. It will show notes from that folder and all subfolders
> 3. It is merged before June 30, 2024 (because I can't put promises out forever!)
> 
> Anyway, I hope all is well and you're still enjoying working on this project :)

if youd be re-willing to make a longer date for this donation, id be happy to add 25$ on top making the total 75$

**Comment by sphh** [comment]
@jsixface: Would it be possible to release a special version with this feature enabled (instead of including it in the main release)?

So whoever wants to test it can then download it from the the github site – and knows, that it might not work reliably. This could also give you some early warnings what is working and what is not working.

**Comment by RustoMCSpit** [comment]
https://github.com/quillpad/quillpad/tree/new-local-storage

theres a new branch that seems to be active, hopefully local storage will be soon!

**Comment by RustoMCSpit** [comment]
by the way, these quillpad files also contain audio, so theyd have to be stored too

**Comment by RustoMCSpit** [comment]
> @jsixface: Would it be possible to release a special version with this feature enabled (instead of including it in the main release)?
> So whoever wants to test it can then download it from the the github site – and knows, that it might not work reliably. This could also give you some early warnings what is working and what is not working.

this has been suggested many times, i dont think @jsixface wants it though!

**Comment by RustoMCSpit** [comment]
> https://github.com/quillpad/quillpad/tree/new-local-storage
> 
> theres a new branch that seems to be active, hopefully local storage will be soon!

just keep your eye on this @sphh

**Comment by sphh** [comment]
> this has been suggested many times, i dont think @jsixface wants it though!

Yes, but all suggested to publish it anyway. My suggestion is a bit different: Compile and publish it in such a way, that it can only be installed manually. Also ‘alpha’, ‘testing’ etc can be put into the apk's file name, which should make it pretty clear, that this is only for the adventurously minded people!

> just keep your eye on this @sphh

The last commit is now 2 months old. Is this another dead end?

**Comment by ConstantMan** [comment]
Is there any progress ?

**Comment by RustoMCSpit** [comment]
> The last commit is now 2 months old. Is this another dead end?
> Is there any progress ?

look, the realistic only solution at this point is to offer to make a donation.  ive offered 25 dollars here. hopefully we offer jsixface or some other dev enough for this. thats it realistically. https://github.com/quillpad/quillpad/issues/26#issuecomment-2267564815

**Comment by sphh** [comment]
@RustoMCSpit, is this way forward agreed by @jsixface? I cannot see him reacting to this offer in this thread.

**Comment by RustoMCSpit** [comment]
> @RustoMCSpit, is this way forward agreed by @jsixface? I cannot see him reacting to this offer in this thread.

no, thats just the only logical way. if we pay them and make it worth their while, then theyd be able to make time for it. thats how all FOSS things work.

**Comment by jsixface** [comment]
Hey everyone, I apologize for not responding to this thread earlier.

I’m not interested in compensation for feature development. As mentioned in other threads, I’m open to reviewing pull requests and merging them.

Here’s the current status of this feature:

I’ve completed about two-thirds of the work, but I’ve been unable to dedicate much time to it recently due to personal commitments. The work I’ve done so far is in a separate branch, so if anyone wants to continue, they can take it from there.

Additionally, I no longer have an Android phone, so I can’t test the feature on a real device. Since this is a significant overhaul of the sync backend, I want to ensure that people’s notes are not lost in case of any issues.

**Comment by sphh** [comment]
Thanks @jsixface for clarifying your view! I do hope, that you will find the time again (and an Android phone) to keep up that good work!

**Comment by RustoMCSpit** [comment]
> Additionally, I no longer have an Android phone, so I can’t test the feature on a real device. Since this is a significant overhaul of the sync backend, I want to ensure that people’s notes are not lost in case of any issues.

i know you said you dont want compensation but im willing to contribute to a gofundme style thing chipping in money for you to be able to get a second hand one if that'd be cool with you. also thank you for replying!

**Comment by RustoMCSpit** [comment]
> I’ve completed about two-thirds of the work, but I’ve been unable to dedicate much time to it recently due to personal commitments. The work I’ve done so far is in a separate branch, so if anyone wants to continue, they can take it from there.

@jsixface 

could you give us a general step by step guide on what needs to be done thats left to do

**Comment by RustoMCSpit** [comment]
> > I’ve completed about two-thirds of the work, but I’ve been unable to dedicate much time to it recently due to personal commitments. The work I’ve done so far is in a separate branch, so if anyone wants to continue, they can take it from there.
> 
> [@jsixface](https://github.com/jsixface)
> 
> could you give us a general breakdown / step by step guide on what needs to be done thats left to do

@jsixface Sorry to bother you but any update?

**Comment by jsixface** [comment]
I did make some progress. The change I'm planning to make is in two stages - Koin migration and local storage. I can produce dev builds which I would like to be alpha tested. The first stage is ready to be tested and can be downloaded from the artifact section of the Github Action (https://github.com/quillpad/quillpad/actions/runs/14606345961). I will post here when the dev build for local storage is ready.

**Comment by RustoMCSpit** [comment]
> I will post here when the dev build for local storage is ready.

Sounds good, the update before you release this update you should force the user to make a backup or they can't access the app.

**Comment by RustoMCSpit** [comment]
Anyways, post the dev build, I'm sure many people here will bugtest for you.

**Comment by madsamjp** [comment]
I'm also eagerly awaiting this feature. I'd love to start using this app for taking notes on my phone, but don't want my notes to be trapped inside the app. I need to sync my notes with my laptop.

**Comment by mezhgano** [comment]
I think this is one of the most important features with highest priority

**Comment by ConstantMan** [comment]
> I think this is one of the most important feature with highest priority

Exactly it's literally the most commented issue

**Comment by jsixface** [comment]
There is a development version ready for testing. You can download it from the [GitHub Actions](https://github.com/quillpad/quillpad/actions/runs/15671439704/artifacts/3332861524). **PLEASE BACKUP YOUR NOTES BEFORE PROCEEDING**.
You can go to settings -> Sync settings -> Choose file storage as a backend and Choose a folder. This can technically be any folder, even the ones provided by other storage apps.

P.S: This version will not map notebooks to sub folders. Nextcloud category to notebook mapping is also broken.

Let me know the feedback in this [discussion thread](https://github.com/quillpad/quillpad/discussions/376)

**Comment by jsixface** [comment]
This is available in the latest version `v1.5`

**Comment by RustoMCSpit** [comment]
when you export files to markdown, the date of the file being created should remain the same. all my files now have the exact same date attached to them.

**Comment by regs01** [comment]
There is a issue with synching - titles won't sync at all.
Created separate #563

**Comment by regs01** [comment]
The feature is still severely broken.
#322 is the main issue ticket
